File: backend/app/services/entry_service.py
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy.orm import Session

from app.models.entry import Entry


def get_active_entry(db: Session, user_id: int) -> Entry | None:
    return db.query(Entry).filter(
        Entry.user_id == user_id,
        Entry.ended_at.is_(None),
    ).first()


def start_entry(db: Session, user_id: int, label: str) -> Entry:
    active = get_active_entry(db, user_id)
    now = datetime.now(timezone.utc)
    if active:
        active.ended_at = now
        db.flush()
    entry = Entry(user_id=user_id, label=label, started_at=now)
    db.add(entry)
    db.commit()
    db.refresh(entry)
    return entry


def stop_entry(db: Session, user_id: int) -> Entry | None:
    active = get_active_entry(db, user_id)
    if active is None:
        return None
    active.ended_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(active)
    return active


def get_entries_by_date(db: Session, user_id: int, date: str) -> list[Entry]:
    from datetime import datetime as dt_mod
    from app.config import TIMEZONE
    import zoneinfo

    tz = zoneinfo.ZoneInfo(TIMEZONE)
    start = dt_mod.strptime(date, "%Y-%m-%d").replace(tzinfo=tz)
    end = dt_mod(start.year, start.month, start.day, 23, 59, 59, tzinfo=tz)
    return db.query(Entry).filter(
        Entry.user_id == user_id,
        Entry.started_at >= start,
        Entry.started_at <= end,
    ).order_by(Entry.started_at).all()


def get_entries_by_range(db: Session, user_id: int, from_dt: datetime, to_dt: datetime) -> list[Entry]:
    return db.query(Entry).filter(
        Entry.user_id == user_id,
        Entry.started_at >= from_dt,
        Entry.started_at <= to_dt,
    ).order_by(Entry.started_at).all()


def update_entry(db: Session, entry_id: int, user_id: int, data: dict) -> Entry | None:
    entry = db.query(Entry).filter(Entry.id == entry_id, Entry.user_id == user_id).first()
    if entry is None:
        return None
    for key, value in data.items():
        if value is not None:
            setattr(entry, key, value)
    db.commit()
    db.refresh(entry)
    return entry


def delete_entry(db: Session, entry_id: int, user_id: int) -> bool:
    entry = db.query(Entry).filter(Entry.id == entry_id, Entry.user_id == user_id).first()
    if entry is None:
        return False
    db.delete(entry)
    db.commit()
    return True

File: backend/src/services/entry-service.ts
import { getDb, saveDb } from "../db.js";

export interface EntryRow {
  id: number;
  user_id: number;
  label: string;
  started_at: string;
  ended_at: string | null;
  created_at: string;
}

function query(sql: string, params: any[] = []): EntryRow[] {
  const db = getDb();
  const stmt = db.prepare(sql);
  stmt.bind(params);
  const rows: EntryRow[] = [];
  while (stmt.step()) {
    rows.push(stmt.getAsObject() as unknown as EntryRow);
  }
  stmt.free();
  return rows;
}

function run(sql: string, params: any[] = []) {
  const db = getDb();
  db.run(sql, params);
  saveDb();
}

function getEntryById(entryId: number, userId: number): EntryRow | null {
  const rows = query("SELECT * FROM entries WHERE id = ? AND user_id = ?", [entryId, userId]);
  return rows[0] || null;
}

export function getActiveEntry(userId: number): EntryRow | null {
  const rows = query("SELECT * FROM entries WHERE user_id = ? AND ended_at IS NULL ORDER BY started_at DESC LIMIT 1", [userId]);
  return rows[0] || null;
}

export function startEntry(userId: number, label: string): EntryRow {
  const active = getActiveEntry(userId);
  const now = new Date().toISOString();
  if (active) {
    run("UPDATE entries SET ended_at = ? WHERE id = ?", [now, active.id]);
  }
  run("INSERT INTO entries (user_id, label, started_at) VALUES (?, ?, ?)", [userId, label, now]);
  const db = getDb();
  const stmt = db.prepare("SELECT * FROM entries ORDER BY id DESC LIMIT 1");
  stmt.bind([]);
  stmt.step();
  const row = stmt.getAsObject() as unknown as EntryRow;
  stmt.free();
  return row;
}

export function stopEntry(userId: number): EntryRow | null {
  const active = getActiveEntry(userId);
  if (!active) return null;
  const now = new Date().toISOString();
  run("UPDATE entries SET ended_at = ? WHERE id = ?", [now, active.id]);
  const rows = query("SELECT * FROM entries WHERE id = ?", [active.id]);
  return rows[0] || null;
}

export function getEntriesByDate(userId: number, date: string): EntryRow[] {
  const start = new Date(`${date}T00:00:00`).toISOString();
  const end = new Date(`${date}T23:59:59`).toISOString();
  return query("SELECT * FROM entries WHERE user_id = ? AND started_at >= ? AND started_at <= ? ORDER BY started_at", [userId, start, end]);
}

export function getEntriesByRange(userId: number, fromDt: string, toDt: string): EntryRow[] {
  return query("SELECT * FROM entries WHERE user_id = ? AND started_at >= ? AND started_at <= ? ORDER BY started_at", [userId, fromDt, toDt]);
}

export function updateEntry(entryId: number, userId: number, data: Record<string, unknown>): EntryRow | null {
  const entry = getEntryById(entryId, userId);
  if (!entry) return null;
  const fields = Object.keys(data).filter((k) => data[k] !== undefined);
  if (fields.length === 0) return entry;
  const setClause = fields.map((f) => `${f} = ?`).join(", ");
  const values = fields.map((f) => data[f]);
  run(`UPDATE entries SET ${setClause} WHERE id = ?`, [...values, entryId]);
  return getEntryById(entryId, userId);
}

export function deleteEntry(entryId: number, userId: number): boolean {
  const db = getDb();
  db.run("DELETE FROM entries WHERE id = ? AND user_id = ?", [entryId, userId]);
  saveDb();
  return db.getRowsModified() > 0;
}